`Communication`: Add profile picture support

PaRangger opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.


#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Currently, the readability of the communication module is lacking in terms of being able to distinguish who wrote a message. This is due to the only visual que being the username, which are hard to differentiate from another without close inspection.

### Description
<!-- Describe your changes in detail -->
I added support for profile pictures in the communication module. This was done by adding a possibility to upload a picture in the user settings. The uploaded picture will be displayed when you write a message or answer in a thread. As a default picture, a user will be assigned a random color hue, which will be taken as the background of the default picture, and their initials will be displayed in white bold text. For example "Tom Martin Muster" will be "TM". This change also required slight rework in the displaying of the post headers. The user role is not displayed as a colorful badge anymore, but simply put below the username to not overload the colors.

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Course with communication enabled
- 1 Student/Tutor/Instructor

1. Log in to Artemis
2. Go to the course
3. Go to the communication tab
4. Switch to any channel the user can write in
5. Post a new message
6. Verify the displaying of the default picture
7. On the top right, click the username and go to settings
8. Press the "Add a profile picture" button
9. Select a picture and crop accordingly
10. Revisit the communication tab and verify if the image is shown in the messages

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked
> Click on the badges to get to the test servers.

[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)](https://artemis-test1.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)](https://artemis-test2.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)](https://artemis-test3.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)](https://artemis-test4.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)](https://artemis-test5.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)](https://artemis-test6.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)](https://artemis-test9.artemis.cit.tum.de)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

<img width="1286" alt="Bildschirmfoto 2024-09-05 um 14 32 14" src="https://github.com/user-attachments/assets/6ec23c7e-53d3-48b6-a83b-728b8b41c34c">

<img width="886" alt="Bildschirmfoto 2024-09-03 um 21 36 00" src="https://github.com/user-attachments/assets/7ce3ed92-a1ec-457a-b435-afddaebd3578">

<img width="506" alt="Bildschirmfoto 2024-09-03 um 21 36 28" src="https://github.com/user-attachments/assets/86d30f18-678e-4f07-a5e5-5c7f947fd79a">

<img width="884" alt="Bildschirmfoto 2024-09-03 um 21 36 37" src="https://github.com/user-attachments/assets/73a6f279-b0de-4bef-9143-5beb6d00fad0">


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Users can now upload, display, and delete their profile pictures.
	- New API endpoints for managing user profile images have been added.
	- Enhanced user interface for managing profile pictures in account settings.
	- Users can now view author profile pictures in posts.
	- Improved visual representation of user roles with icons.

- **Bug Fixes**
	- Fixed issues with displaying default profile pictures when none are set.

- **Documentation**
	- Updated localization files to include error messages for profile picture uploads.

- **Tests**
	- Added tests for updating and deleting user profile pictures to ensure functionality works as expected.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PaRangger commented on src/main/webapp/app/course/manage/course-update.component.ts:
```diff
@@ -639,7 +639,7 @@ export class CourseUpdateComponent implements OnInit {
     openCropper(): void {
         const modalRef = this.modalService.open(ImageCropperModalComponent, { size: 'm' });
-        modalRef.componentInstance.courseImageUploadFile = this.courseImageUploadFile;
+        modalRef.componentInstance.uploadFile = this.courseImageUploadFile;
```
Checked all uses, could not find any other place where it should be renamed

anian03 commented:
Just tested it and it worked as expected :)
One question: Is a gravatar integration possible/planned? I think it would be nice if we could make the user's gravatar picture the default (instead of none), just like Slack or Zulip do

anian03 reviewed (changes requested):
Two suggestions:
- I think it would be nice if you could click on your own profile picture in a conversation to have an option to change it there. Users might not know to go to settings to do that.
- And should we maybe also display the profile picture here? <img src="https://github.com/user-attachments/assets/6c82d8b9-9229-4251-8792-ea9b3c9114a4" width="300">

bassner commented:
Gravatar should definitely be implemented, ig. in a follow-up

PaRangger commented:
> Just tested it and it worked as expected :) One question: Is a gravatar integration possible/planned? I think it would be nice if we could make the user's gravatar picture the default (instead of none), just like Slack or Zulip do

Good idea! I think we can add this in a standalone PR afterwards.

SimonEntholzer commented on src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:
```diff
@@ -115,6 +126,55 @@ public ResponseEntity<Void> updateUserNotificationDate() {
+     * @param file the image file that is being uploaded
+     * @return the ResponseEntity with status 200 (OK) and with body of current user
+     */
+    @PutMapping("users/profile-picture")
+    @EnforceAtLeastStudent
+    public ResponseEntity<UserDTO> updateProfilePicture(@RequestPart MultipartFile file) throws URISyntaxException {
```
I guess it's not entirely clear where the actual difference is, but most other settings use the `account/....` endpoint. I think also doing that for the profile picture, might align a bit better with the rest of the settings endpoints.

PaRangger replied:
Hey! Yeah, you are right, in my mind the account api was used for mainly managing credentials (username, password, tokens) and the user api was used for metadata. I was a bit on the fence when deciding which one to go for. The decision was mainly influenced by the administration api also being called "admin/users". But I'll change it to /account, I think that makes more sense. 👍🏼

PaRangger replied:
I changed the api route to /api/account/profile-picture and moved the functions to the AccountResource in this [commit](https://github.com/ls1intum/Artemis/pull/9277/commits/d9a77eb44ab7f0ade78c98d3fbb56c58b9c85ad5). 👍🏼

PaRangger commented:
> Two suggestions:
> 
> * I think it would be nice if you could click on your own profile picture in a conversation to have an option to change it there. Users might not know to go to settings to do that.
> * And should we maybe also display the profile picture here? <img alt="" width="300" src="https://private-user-images.githubusercontent.com/98647423/364986530-6c82d8b9-9229-4251-8792-ea9b3c9114a4.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2NDM3NjMsIm5iZiI6MTcyNTY0MzQ2MywicGF0aCI6Ii85ODY0NzQyMy8zNjQ5ODY1MzAtNmM4MmQ4YjktOTIyOS00MjUxLTg3OTItZWE5YjNjOTExNGE0LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNlQxNzI0MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Y2JhNDg0NWQ0ZjMzMDk1ODk1NTIyNTU4ZGExNjBmOGY5YjI1ZjVjNDc5ZTVmNGZhMmRkZjQxNzRiM2IyMGZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wYIo3IOpjkLotIzh7Wap_hDwSPrCMArK7aG18YQyGtA">

I added the picture to the profile and added an option to navigate to settings by clicking the user icon of your own messages. 😊

<img width="257" alt="Bildschirmfoto 2024-09-06 um 08 32 53" src="https://github.com/user-attachments/assets/c0c7c5c7-70bd-4e69-b13f-88bebedae48b">
 
<img width="362" alt="Bildschirmfoto 2024-09-06 um 09 30 49" src="https://github.com/user-attachments/assets/674a2003-b1a5-4aa1-970e-bb2dfbf97012">

asliayk reviewed (approved):
great work :) tested locally and everything works as expected

dfuchss reviewed (commented):
Please ensure that the max size of the images is very low. Keep in mind that multiple 1000 Students are registered to artemis. Also the images shall not remain if a user gets deleted (-> GDPR)

PaRangger commented:
> Please ensure that the max size of the images is very low. Keep in mind that multiple 1000 Students are registered to artemis. Also the images shall not remain if a user gets deleted (-> GDPR)

Good call! I added functionality to remove the image on user deletion and adressed the image size with a max upload size restriction as well as a change of upload format to jpeg (when using the cropper). The cropper itself already is configured to scale the images to a max of 200x200. These changes should guarantee smaller filesizes in general, as well as prevent massive filesizes if anyone uploads an image to the api directly, without using the client.

anian03 reviewed (commented):
Thanks for the changes! Works as expected :)

One more thing I noticed: Sometimes I get this weird black square appearing at the top left corner after/while using the photo cropper. Especially switching between light and dark mode causes this. Is this related to this PR or a general issue with the Photo Cropper?
<img src="https://github.com/user-attachments/assets/940b7657-73bf-4090-a0fc-8d309490908e" width="300">

PaRangger commented:
> Thanks for the changes! Works as expected :)
> 
> One more thing I noticed: Sometimes I get this weird black square appearing at the top left corner after/while using the photo cropper. Especially switching between light and dark mode causes this. Is this related to this PR or a general issue with the Photo Cropper? <img alt="" width="300" src="https://private-user-images.githubusercontent.com/98647423/365408801-940b7657-73bf-4090-a0fc-8d309490908e.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU3NzM1ODgsIm5iZiI6MTcyNTc3MzI4OCwicGF0aCI6Ii85ODY0NzQyMy8zNjU0MDg4MDEtOTQwYjc2NTctNzNiZi00MDkwLWEwZmMtOGQzMDk0OTA5MDhlLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwOFQwNTI4MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NTQ3YmMwOGY2NDEzYjVhMzAxMDQyYTQ3YWQ5ZDk4NGRmYTU4YmJjOGYyNGM0ZWIwNDNmNzQzZDU1Yzg5Mjg2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.gDTmKUL4fM0wEW8Q6rYpPnmfbr06pBTqWpqunthMVOk">

I noticed a similar issue one time but just discarded it. I think it‘s related to the canvas that gets created by the cropper. I don‘t think its related to the changes in the PR, but I have an idea to fix it. As soon as I manage to reproduce it consistently I can create another PR to address this.

SimonEntholzer reviewed (approved):
Code lgtm

edkaya reviewed (approved):
Tested on TS6, works as expected. Really nice feature btw!

BBesrour reviewed (approved):
Tested on ts6, lgtm!

sarpsahinalp reviewed (approved):
Code lgtm

PaRangger commented on src/main/webapp/app/shared/metis/posting-header/posting-header.directive.ts:
```diff
@@ -63,21 +79,83 @@ export abstract class PostingHeaderDirective<T extends Posting> implements OnIni
+     * Returns a pseudo-random numeric value for a given string using a simple hash function.
+     * @param {string} str - The string used for the hash function.
+     */
+    private deterministicRandomValueFromString(str: string): number {
```
Hi, thanks for the feedback!

Sanitization like filename transformation and hiding the path were already implemented by the File Service. I heavily oriented myself by the implementation of uploading course images. I mainly focused on validation of the file in this pr 👍

And concerning the hash function: I did not find a proper pseudo random hash function on the client side. That's why I chose to implement a very simple one for this component to generate a deterministic color hue. The hashCode function is on the server-side, or did I overlook something?
